statefun-sdk/src/main/java/org/apache/flink/statefun/sdk/state/PersistedAppendingBuffer.java

```diff
@@ -0,0 +1,189 @@
+   * Creates a {@link PersistedAppendingBuffer} instance that may be used to access persisted state
+   * managed by the system. Access to the persisted buffer is identified by an unique name and type
+   * of the elements. These may not change across multiple executions of the application.
+   *
+   * @param name the unique name of the persisted buffer state
+   * @param elementType the type of the elements of this {@code PersistedAppendingBuffer}.
+   * @param <E> the type of the elements.
+   * @return a {@code PersistedAppendingBuffer} instance.
+   */
+  public static <E> PersistedAppendingBuffer<E> of(String name, Class<E> elementType) {
+    return new PersistedAppendingBuffer<>(name, elementType, new NonFaultTolerantAccessor<>());
+  }
+
+  /**
+   * Returns the unique name of the persisted buffer.
+   *
+   * @return unique name of the persisted buffer.
+   */
+  public String name() {
+    return name;
+  }
+
+  /**
+   * Returns the type of the persisted buffer elements.
+   *
+   * @return the type of the persisted buffer elements.
+   */
+  public Class<E> elementType() {
+    return elementType;
+  }
+
+  /**
+   * Appends an element to the persisted buffer.
+   *
+   * <p>If {@code null} is passed in, then this method has no effect and the persisted buffer
+   * remains the same.
+   *
+   * @param element the element to add to the persisted buffer.
+   */
+  public void append(@Nullable E element) {
+    if (element != null) {
+      accessor.append(element);
+    }
+  }
+
+  /**
+   * Adds all elements of a list to the persisted buffer.
+   *
+   * <p>If {@code null} or an empty list is passed in, then this method has no effect and the
+   * persisted buffer remains the same.
+   *
+   * @param elements a list of elements to add to the persisted buffer.
+   */
+  public void appendAll(@Nullable List<E> elements) {
+    if (elements != null && !elements.isEmpty()) {
+      accessor.appendAll(elements);
+    }
+  }
+
+  /**
+   * Replace the elements in the persisted buffer with the provided list of elements.
+   *
+   * <p>If an empty list or {@code null} is passed in, this method will have the same effect as
+   * {@link #clear()}.
+   *
+   * @param elements list of elements to replace the elements in the persisted buffer with.
+   */
+  public void replaceWith(@Nullable List<E> elements) {
+    if (elements != null && !elements.isEmpty()) {
+      accessor.replaceWith(elements);
+    } else {
+      accessor.clear();
+    }
+  }
+
+  /**
+   * Gets the elements of the persisted buffer as an {@link Iterable}.
+   *
+   * <p>This may return {@code null} if the buffer is empty or had been cleared (with {@link
+   * #clear()}).
+   *
+   * <p>Note that any modifications to the view does not modify the elements in the persisted
+   * buffer. To modify elements in the persisted buffer, you must use {@link #replaceWith(List)}.
+   *
+   * @return a modifiable view, as an {@link Iterable}, of the elements of the persisted buffer, or
```

Comment:
Is it modifiable or unmodifiable?

Reply:
The view is modifiable, e.g. Iterator.remove() is supported (I got it wrong during our discussion yesterday). But the persisted state will not be modified.

Keep in mind that this is only true for RocksDB and not for heap - 
but my thinking is that if the contract is set such that the view is always a snapshot copy and `replace` needs to be always be used to modify the persisted state, then it would work for both backends.

Reply:
Interesting. Does this mean that using RocksDB we can implement efficient head removal (poping the first element)
If this there case, then I think we should add a pop method, and change the view method to simply items().
If later we would decide to support heap backend, then we can provide a different accessory that supports pop() by calling
list.sublist(1, list.size())

Reply:
Interesting. Does this mean that using RocksDB we can implement efficient head removal (poping the first element)
If this there case, then I think we should add a pop method, and change the view method to simply items().
If later we would decide to support heap backend, then we can provide a different accessory that supports pop() by calling
list.sublist(1, list.size())

Reply:
no, RocksDB can not do efficient head removal.

When I say that the view is modifiable, essentially:
You can treat the view that you get as a copy (underneath its just an `ArrayList`). BUT this is only the case for RocksDB.

Reply:
Got it, then in that case.
Let’s reinforce the contract by wrapping the iterate with a unmodifiable iterable.
Other then that, I’m good for merging this as is!

Reply:
Taking one step back here:

The intention _is_ to have a modifiable copy, so that it'll be easier for users to remove elements from the view, and then call `replaceWith` afterwards with the modified view.

The comment about different between RocksDB and Heap, is that for heap since you don't really get a copy, when modifying the view you already modify the persisted state.

So, the only way to reinforce the contract is to actually always make a copy of the returned `Iterable` in Stateful Functions space, but that's a big performance hit and I'm not sure it's worth doing.

Reply:
Having a modifiable view that support `.remove()`, that doesn't*  remove the actual element from the persisted storage is a confusing API in my mind, and can be easily avoided by wrapping the returned Iterable with an `UnmodifieableIterable` and clearly document this that this is a read only view.

To add to the confusion 🙈  `.remove()` can remove the elements when using the heap state backend (potentially so, depends on rather or not a snapshot is happening concurrently etc')